### ISSUE TITLE
Light bulbs flicker drastically when changing brightness

### ISSUE THREAD

**Issue by damongroenveld** [rudders/homebridge-platform-wemo#43]
Hi, When changing the brightness of lightbulbs they flicker drastically from almost off to brighter than set. Turning off the wemo bulb at the switch and turning it back on often stabilizes it, or helps until it can be turned off via homebridge.
Turning on and off the bulbs works just fine though.
I'm not sure if it is this plugin or MyTouchHome (homekit app).

Changing the brightness in the wemo app works just fine.

**Comment by rjosborne** [comment]
It seems to be an issue caused by the way this app (and others) update brightness - they send every value through e.g. changing from 100 - 90 sends 99,98,97,96,95,94,93,92,91,90.

What should ideally happens is you update the brightness to 90, that gets sent to wemo, wemo tells homebridge its changed to 90, homebridge considers setting the value to 90 and then stops because the homekit value is already set at 90.

Unfortunately, when you send every value through, what happens is:

you update the brightness to 90, which sends 10 updates to homekit for 99 -> 90. The 99 update gets sent to wemo, wemo updates and tells homebridge it's at 99, homebridge then considers setting the value to 99 and because it has by this time been told to change to 98, 97,96,95 etc, it changes back to 99, which then goes to wemo again etc, causing the flickering.

The pull request i've submitted fixes this - there seems to still be some issues with the app as the UI doesn't seem to want to change "cleanly" to the value you choose - it seems to slowly change to your chosen value a few points at a time. However - it fixes the flickering and I can change the brightness perfectly well, so as somebody who has experienced the flickering first hand this is something I can live with.

**Comment by devbobo** [comment]
@rjosborne I suspect that the client app that you are using is displaying a slider to set the value ? As such when you slide the value from 100 to 90, the client app triggers a change in value for every step from 100 to 90.

I assume that if you ask Siri to `Set the lights to 90` that it works fine without any flickering ?

**Comment by rjosborne** [comment]
@devbobo exactly right - Siri works fine. 

#47 fixes the issue (other than the way the slider moves around for about 30 seconds after making the change, but this is just visual with no affect on functionality).

**Comment by driskell** [comment]
Might also be related to the fact that when setting the brightness it doesn't wait for it to complete (via the callback in `wemo-client`) before returning to homebridge. That might cause some weirdness. Really the `setBrightness` needs updating so `cb` is only called when brightness set completes. That might actually resolve the entire problem.

It could be because things move so quickly, the Wemo receives almost simultaneous requests for brightness 99,98,97,96,95,94,93,92 and effectively depending on network and CPU scheduling it isn't necessarily going to process them in order because they arrived so quickly. Could end up processing like:  94,98,99,96,97,95,92 and completely messing up. Maybe improving the `cb` usage would fix it without any need to debounce?

**Comment by driskell** [comment]
I noticed this in #51 as it looks like it could cause error bubble in iOS 10 home as some similar thing happened to be with Sonos switches when I was implementing new features.

**Comment by driskell** [comment]
Might be worth testing https://github.com/rudders/homebridge-platform-wemo/pull/53 ? It fixes the callback things.

**Comment by devbobo** [comment]
I've bought some bulbs and spent some time refactoring my `WemoLinkAccessory`. I believe that I have resolved the major of issues...anyone interested in testing this ?

**Comment by devbobo** [comment]
@rjosborne check out version `1.1.1`, I've rewritten the `setBrightness` function without the use of `setTimeout`... interested to hear your feedback on performance.

**Comment by hauweit** [comment]
@devbobo I have Wemo bulbs as well, so far it's working pretty solidly (changing brightness) for me on `1.1.2`, any specific behaviour I should test? I'm using the iOS 10 Home app.

**Comment by devbobo** [comment]
@hauweit i don't believe so, I think it was pretty obvious when it was happening previously. Alternating between high/low brightnesses.

**Comment by hauweit** [comment]
@devbobo No worries, I certainly noticed the flickering prior to your refactor. I just worked around it by asking Siri to set specific brightness instead of using sliders in apps.

**Comment by rjosborne** [comment]
@devbobo It doesn't seem to be working quite right here - I'm finding it's noticeably slower to reach the intended value and quite regularly does it in "steps".

e.g:

- I started at 33%
- Using the slider, slide up to 100%
- UI shows 100%
- Light brightens, then brightens further a second time
- UI drops to 58%, then returns to 100%

Although the previous version had the 0.1 second delay before applying the change, this version still seems slower to reach the final value - especially with the "stepping" behaviour I'm seeing.

This is seen with the apple home app. I've also tried it with MyTouchHome which seems even worse, taking 10 or more "steps" at times to reach the final value, with each step separated by a second or two.
